optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/inverserelation/SingletonListInverseVariableListener.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.domain.variable.inverserelation;
+
+import java.util.List;
+import java.util.Objects;
+
+import org.optaplanner.core.api.domain.variable.VariableListener;
+import org.optaplanner.core.api.score.director.ScoreDirector;
+import org.optaplanner.core.impl.domain.variable.descriptor.ListVariableDescriptor;
+import org.optaplanner.core.impl.score.director.InnerScoreDirector;
+
+public class SingletonListInverseVariableListener<Solution_>
+        implements VariableListener<Solution_, Object>, SingletonInverseVariableSupply {
+
+    protected final InverseRelationShadowVariableDescriptor<Solution_> shadowVariableDescriptor;
+    protected final ListVariableDescriptor<Solution_> sourceListVariableDescriptor;
+
+    public SingletonListInverseVariableListener(
+            InverseRelationShadowVariableDescriptor<Solution_> shadowVariableDescriptor,
+            ListVariableDescriptor<Solution_> sourceListVariableDescriptor) {
+        this.shadowVariableDescriptor = shadowVariableDescriptor;
+        this.sourceListVariableDescriptor = sourceListVariableDescriptor;
+    }
+
+    @Override
+    public void beforeEntityAdded(ScoreDirector<Solution_> scoreDirector, Object entity) {
+        // Do nothing
+    }
+
+    @Override
+    public void afterEntityAdded(ScoreDirector<Solution_> scoreDirector, Object entity) {
+        insert((InnerScoreDirector<Solution_, ?>) scoreDirector, entity);
+    }
+
+    @Override
+    public void beforeVariableChanged(ScoreDirector<Solution_> scoreDirector, Object entity) {
+        // TODO maybe set all indexes to null
+        // When we support over-constrained planning (e.g. there is not enough Employees to assign all tasks), then
+        // an unassigned planning value should have a null index.
+    }
+
+    @Override
+    public void afterVariableChanged(ScoreDirector<Solution_> scoreDirector, Object entity) {
+        insert((InnerScoreDirector<Solution_, ?>) scoreDirector, entity);
+    }
+
+    @Override
+    public void beforeEntityRemoved(ScoreDirector<Solution_> scoreDirector, Object entity) {
+        // TODO maybe set all indexes to null
```

Comment:
yes, for CH needed

Reply:
if employee is removed, all its task.employee = null

Reply:
In basic vars, if Shift.employee has an inverse variable Employee.shiftList,
and a PFC removes a shift S1, does it get removed from Ann.shiftList [S1, S2] ?

Reply:
These Added/remove methods exists for:
1) Moves that add/remove shifts etc to cover a certain staffing demand because the solver can create/remove more entities (in inner solver, **never used so far**)
2) PFC that add/remove shifts etc because the problem changed(not in inner solver)
=> impl must definitly work wel with 2), but 1) is nice to have theoretically

Reply:
> In basic vars, if Shift.employee has an inverse variable Employee.shiftList,
> and a PFC removes a shift S1, does it get removed from Ann.shiftList [S1, S2] ?

Yes.

Reply:
Done.